modules/FindFreetype.cmake

```diff
@@ -159,6 +185,59 @@ find_package_handle_standard_args(
     FREETYPE_VERSION_STRING
 )
 
+if(FREETYPE_FOUND)
+  # Determine if Freetype was built as static or shared
+  set(FREETYPE_SHARED OFF)
+  foreach(_path ${FREETYPE_LIBRARIES})
+    string(FIND ${_path} "/" _result REVERSE)
+    string(SUBSTRING ${_path} ${_result} -1 _name)
+    foreach(_hint dll so dylib)
+      string(FIND ${_name} ${_hint} _result)
+      if(NOT _result EQUAL -1)
+        set(FREETYPE_SHARED ON)
+        break()
+      endif()
+    endforeach()
+  endforeach()
+
+  # If Freetype was built static, we need to link its optional dependencies
+  if(NOT FREETYPE_SHARED)
+    # We can read which dependencies Freetype was built with from ftoption.h
+    file(READ "${FREETYPE_INCLUDE_DIR_freetype2}/freetype/config/ftoption.h" _ftoption_h)
+    foreach(_dep ZLIB BZip2 PNG HarfBuzz)
+      string(TOUPPER ${_dep} _DEP)
+      set(_require_dep OFF)
+
+      # Use regex to check if a depdency is needed
+      # Example of when a dependency was used to build Freetype:
+      # #define FT_CONFIG_OPTION_USE_BZIP2
+      # Example of when a dependency was not used to build Freetype:
+      # /* #define FT_CONFIG_OPTION_USE_HARFBUZZ */
+      # We want to match the first and not the second
+      set(_match_before "\n[ \t]*\\#define[ \t]*FT_CONFIG_OPTION")
+      set(_match_after "[ \t]*\n")
+      string(REGEX MATCH "${_match_before}_USE_${_DEP}${_match_after}" _use_dep "${_ftoption_h}")
+      if("${_dep}" STREQUAL "ZLIB")
+        # Special handling
+        string(REGEX MATCH "${_match_before}_SYSTEM_${_DEP}${_match_after}" _sys_dep "${_ftoption_h}")
+        if(NOT "${_use_dep}" STREQUAL "" AND NOT "${_sys_dep}" STREQUAL "")
+          set(_require_dep ON)
+        endif()
+      else()
+        if(NOT "${_use_dep}" STREQUAL "")
+          set(_require_dep ON)
+        endif()
+      endif()
+
+      # Require the dependency only if Freetype was built with it
+      if(_require_dep)
+        find_package(${_dep} QUIET REQUIRED)
+        list(APPEND FREETYPE_DEPENDENCY_LIBRARIES ${${_DEP}_LIBRARY} ${${_DEP}_LIBRARIES})
```

Comment:
Some of the find modules only use one or the other, some use both. I'm just bulletproofing so I don't have to special case them. E.g. There is only `ZLIB_LIBRARIES` and only `HARFBUZZ_LIBRARY`. We could change the latter case, though.